[pytorch][cupti profiler 6/n] Changes to configure Kineto cupti profiler from pytorch profiler interface

briancoutinho opened the pull request:
Summary:
Kineto introduced a new profiler to read performance counters from NVIDIA GPUs (CUPTI Range Profiler API)
Here we are adding support to configure this Kineto range profiler mode

Example
```
    with torch.profiler.profile(
        activities=[ProfilerActivity.CUDA],
        record_shapes=True,
        on_trace_ready=trace_handler,
        experimental_config=torch.profiler.ExperimentalConfig(
            profiler_metrics=[
                "kineto__tensor_core_insts",
                "dram__bytes_read.sum",
                "dram__bytes_write.sum"],
            profiler_measure_per_kernel=False),
    ) as prof:
        res = train_batch(modeldef)
        prof.step()
```

## Details
* Introduce a new structure `ExperimentalConfig` so users can configure Kineto specific options, keeps profiler API consistent.
* Populate configuration options for Kineto.

Test Plan: CI and tested on resnet50

Differential Revision: D34489487

albanD commented on torch/autograd/__init__.py:
```diff
@@ -303,7 +303,7 @@ def variable(*args, **kwargs):
 
 # Import all native method/classes
-from torch._C._autograd import (DeviceType, ProfilerActivity, ProfilerState, ProfilerConfig, ProfilerEvent,
+from torch._C._autograd import (DeviceType, ProfilerActivity, ProfilerState, ProfilerConfig, ProfilerEvent, ExperimentalConfig,
```
Why do you expose this both on autograd, autograd.profiler and profiler? I would expect that this appears only once on the public API no?

briancoutinho replied:
yea the way it is done all of these C types are first exposed in autograd/__init__.py 
then you see them getting imported in torch/profiler, the code is similar for ProfilerActivity
```
from torch.autograd import ProfilerActivity, kineto_available, ExperimentalConfig
```

briancoutinho replied:
sorry let me clarify my comment, you are right that in should be exposed only in public API. I think I followed the current way this is done :

albanD replied:
> As a result you have to import the C types into autograd init.py
torch/autograd/__init__.py

This is not true AFAIK you can import it anywhere you want!

briancoutinho replied:
Oh I didn't know that, let me try to import the csrc autograd.profiler and profiler

briancoutinho replied:
@albanD please let me know if this looks better, thanks.

albanD commented on torch/_C/_autograd.pyi:
```diff
@@ -35,6 +35,14 @@ class DeviceType(Enum):
     Metal = ...
     ...
 
+class _ExperimentalConfig:
```
These are no needed anymore right?

briancoutinho replied:
actually they are because in the same file we have ProfilerConfig and this is an argument to the config 
```
        with_modules: bool,
        experimental_config: _ExperimentalConfig,  <<<<
```

albanD reviewed (commented):
new binding looks good!
Thanks for the update.

briancoutinho commented on torch/_C/_autograd.pyi:
```diff
@@ -43,7 +51,8 @@ class ProfilerConfig:
         with_flops: bool,
-        with_modules: bool
+        with_modules: bool,
+        experimental_config: _ExperimentalConfig,
```
@albanD  here is the usages of the object.

albanD commented:
Thanks for taking the time to clean this up!
The autograd binding part looks good to me!

briancoutinho commented:
Okay CI failure is unrelated to this diff : it fails on this binding
```
AssertionError: False is not true : Following new APIs ( displayed in the form (module, element, element module) ) were added that do not meet our guidelines for public API Please review  https://docs.google.com/document/d/10yx2-4gs0gTMOimVS403MnoAWkqitS8TUHX73PN8EjE/edit?pli=1# for more information:
('torch.ao.quantization.fx.backend_config_utils', 'get_quantize_handler_cls', 'torch.ao.quantization.fx.backend_config_utils')
```

albanD commented:
Yes, rebasing on top of master should fix this.

Commit:
[pytorch][cupti profiler 6/n] Changes to configure Kineto cupti profiler from pytorch profiler interface (#75616)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/75616

Kineto introduced a new profiler to read performance counters from NVIDIA GPUs (CUPTI Range Profiler API)
Here we are adding support to configure this Kineto range profiler mode

Example
```
    with torch.profiler.profile(
        activities=[ProfilerActivity.CUDA],
        record_shapes=True,
        on_trace_ready=trace_handler,
        experimental_config=torch.profiler._ExperimentalConfig(
            profiler_metrics=[
                "kineto__tensor_core_insts",
                "dram__bytes_read.sum",
                "dram__bytes_write.sum"],
            profiler_measure_per_kernel=False),
    ) as prof:
        res = train_batch(modeldef)
        prof.step()
```

## Details
* Introduce a new structure `KinetoProfilerConfig` so users can configure Kineto specific options, keeps profiler API consistent.
* Populate configuration options for Kineto.

Test Plan: CI and tested on resnet50

Differential Revision: D34489487

fbshipit-source-id: d5ad6dc508fd13c297e8083482c5ea89ccfb3abd
